perf/uncore_libpfm.go

```diff
@@ -157,56 +160,101 @@ func NewUncoreCollector(cgroupPath string, events PerfEvents, topology []info.No
 }
 
 func (c *uncoreCollector) setup(events PerfEvents, devicesPath string) error {
-	var err error
 	readUncorePMUs, err := getUncorePMUs(devicesPath)
 	if err != nil {
 		return err
 	}
 
-	// Maping from event name, pmu type, cpu.
-	c.cpuFiles = make(map[string]map[string]map[int]readerCloser)
+	c.cpuFiles = make(map[int]map[string]group)
 	c.events = events.Uncore.Events
 	c.eventToCustomEvent = parseUncoreEvents(events.Uncore)
 	c.cpuFilesLock.Lock()
 	defer c.cpuFilesLock.Unlock()
 
-	for _, group := range c.events {
-		if len(group) > 1 {
-			klog.Warning("grouping uncore perf events is not supported!")
-			continue
+	for i, group := range c.events {
+		// Check what PMUs are needed.
+		groupPMUs, err := parsePMUs(group, readUncorePMUs, c.eventToCustomEvent)
+		if err != nil {
+			return err
 		}
 
-		eventName, pmuPrefix := parseEventName(string(group[0]))
+		err = checkGroup(group, groupPMUs)
+		if err != nil {
+			return err
+		}
 
-		var err error
-		customEvent, ok := c.eventToCustomEvent[group[0]]
-		if ok {
-			if customEvent.Type != 0 {
-				pmuPrefix = uncorePMUPrefix
+		// CPUs file descriptors of group leader needed for perf_event_open.
```

Comment:
These are uncore metrics. Why they are CPU dependent?

Reply:
Because of `perf_event_open` arguments. There is an abstraction that one CPU corresponds to one socket. 

For example CPU range in `/sys/device/uncore_imc_0/cpumask` of 2 socket machine:

```
$ cat /sys/devices/uncore_imc_0/cpumask 
0-1
```